src/osp/Active/SysVehicle.h

```diff
@@ -111,16 +112,70 @@ class SysVehicle : public IDynamicSystem, public IActivator
+    struct MachineDef
+    {
+        ActiveEnt m_machineOwner;
+        std::vector<unsigned> const& m_machineIndices;
+    };
+
+    /**
+     * Add machines to the specified entity
+     *
+     * Receives the master arrays of prototype and blueprint machines from a
+     * prototype/blueprint part, as well as an array of indices that describe
+     * which of the machines from the master list are to be instantiated for
+     * the specified entity.
+     * @param partEnt [in] The part entity which owns the ACompMachines component
+     * @param entity [in] The target entity to receive the machines
+     * @param protoMachines [in] The master array of prototype machines
+     * @param blueprintMachines [in] The master array of machine configs
+     * @param machineIndices [in] The indices of the machines to add to entity
+     */
+    void add_machines_to_object(ActiveEnt partEnt, ActiveEnt entity,
+        std::vector<PrototypeMachine> const& protoMachines,
+        std::vector<BlueprintMachine> const& blueprintMachines,
+        std::vector<unsigned> const& machineIndices);
+
+    /**
+     * Instantiate all part machines
+     *
+     * Machine instantiation requires the part's hierarchy to already exist
+     * in case a sub-object needs information about its peers. Since this means
+     * machines can't be instantiated alongside their associated object, the
+     * association between entities and prototype machines is captured, then used
+     * to call this function after all part children exist to instance all the
+     * machines at once.
+     * 
+     * Effectively just calls add_machines_to_object() for each object in the
+     * part.
+     * 
+     * @param partEnt [in] The root entity of the part
+     * @param machineMapping [in] The mapping from objects to their machines
+     * @param part [in] The prototype part being created
+     * @param partBP [in] The blueprint configs of the part being created
+     */
+    void part_instantiate_machines(ActiveEnt partEnt,
+        std::vector<MachineDef> machineMapping,
+        PrototypePart& part, BlueprintPart& partBP);
+
     /**
      * Create a Physical Part from a PrototypePart and put it in the world
-     * @param part the part to instantiate
-     * @param rootParent Entity to put part into
+     * @param part [in] The part prototype to instantiate
+     * @param blueprint [in] Unique part configuration data
+     * @param rootParent [in] Entity to put part into
+     * @param machineDefinitions [out] List of part machines for later creation
      * @return Pointer to object created
      */
-    ActiveEnt part_instantiate(PrototypePart& part, ActiveEnt rootParent);
-
-    // Handle deleted parts and separations
-    void update_vehicle_modification(ActiveScene& rScene);
+    ActiveEnt part_instantiate(PrototypePart& part, BlueprintPart& blueprint,
+        ActiveEnt rootParent, std::vector<MachineDef>& machineDefinitions);
```

Comment:
I meant to have this return a tuple and use structured bindings instead of the `machineDefinitions` output parameter; I'll do it tomorrow